### PR TITLE
Implement Planar Multigraph with HalfEdge Data Structure

### PR DESCRIPTION
**Pull request by exAClior**

### CODE REVIEW

**Comment by exAClior** [comment]
We follow the API from this [library](https://doc.cgal.org/latest/Polyhedron/classCGAL_1_1Polyhedron__3.html#a2b17d7bd2045397167b00616f3b4d622)

**Review by ChenZhao44 — COMMENTED** [review]
Looks great in general. We can merge it once the tests of `ZXW` are fixed.

File: test/planar_multigraph.jl [hunk 1]
```diff
@@ -0,0 +1,724 @@
...
+        5,
+        10,
+        1,
+    )
+    pmg2f1 = copy(pmg1)
+    @test split_vertex!(pmg2f1, 4, 1) == 9
+    @test pmg2f1 == pmg2
+end
+
+@testset "Make hole" begin
```
[the remark this answers is not in the document]
Comment (exAClior) [reply test/planar_multigraph.jl:432]: Thank you very much for reviewing the code! I will make changes accordingly and explain during our next meetup. I also had some concerns about making every hole facet 0. Your suggestion of seems good and easy. I will implement it.

File: src/ZXCalculus.jl [hunk 2]
```diff
@@ -6,34 +6,59 @@ using YaoHIR.IntrinsicOperation
...
+    full_reduction,
+    circuit_extraction,
+    phase_teleportation
 export substitute_variables!, expval_circ!, stack_zxwd!, concat!
 
+export spider_type, rem_spiders!, scalar, rem_spider!, phase, spiders
+
+import Graphs: has_vertex
+
+include("adts.jl")
```
[the remark this answers is not in the document]
Comment (exAClior) [reply src/ZXCalculus.jl:48]: Sounds good, will do that!

**Pull request merged**